hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterSnapshot.java

```diff
@@ -157,6 +158,24 @@ public void renameSnapshot(String snapshotRoot, String oldSnapshotName,
     return RouterRpcServer.merge(ret, SnapshottableDirectoryStatus.class);
   }
 
+  public SnapshotStatus[] getSnapshotListing(String snapshotRoot)
+      throws IOException {
+    rpcServer.checkOperation(NameNode.OperationCategory.READ);
+    final List<RemoteLocation> locations =
+        rpcServer.getLocationsForPath(snapshotRoot, true, false);
+    RemoteMethod remoteMethod = new RemoteMethod("getSnapshotListing",
+        new Class<?>[]{String.class},
+        new RemoteParam());
+    if (rpcServer.isInvokeConcurrent(snapshotRoot)) {
+      Map<RemoteLocation, SnapshotStatus[]> ret = rpcClient.invokeConcurrent(
+          locations, remoteMethod, true, false, SnapshotStatus[].class);
+      return ret.values().iterator().next();
+    } else {
+      return rpcClient.invokeSequential(
+          locations, remoteMethod, SnapshotStatus[].class, null);
```

Comment:
Here the `SnapshotStatus` contains `parentpath` as well, the path will be the one wrt the actual namespace not with respect to mount table. You need to replace the `parentpath` with path corresponding to the mount entry.
if mount entry is : `/mnt -> /dir` in `ns0`, then if you trigger call on `ns0` the `path` would be something like `/dir/sub..` that you need to change to `/mnt/sub..`
In case of `InvokeConcurrent` you would be able to get `Src` and `Dst` from the `ret` and maybe something like this may work  -
```
      response = ret.values().iterator().next();
      String src = ret.keySet().iterator().next().getSrc();
      String dst = ret.keySet().iterator().next().getDest();
      for (SnapshotStatus s : response) {
        String mountPath =
            new String(s.getParentFullPath()).replaceFirst(dst, src);
        s.setParentFullPath(mountPath.getBytes());
      }
```

For the `invokeSequential` one you won't be having the detail on which `location` did the call got success, For that you have to get it returned back from `RouterRpcClient#L858`, I guess to get the location returned you would require a new `InvokeSequential` method which returns the `location` as well, may be can refactor and reuse this one...

This problem would be there I think in `getSnapshottableDirListing()` as well. If you want, you can put a `TODO` and handle that in a separate follow up jira.